libgadget/uvbg.c

```diff
@@ -0,0 +1,971 @@
+    else {
+        // calculate midpoint to cut set in half
+        int imid = imin + ((imax - imin) / 2);
+        void* arr_val = (void*)(((char*)arr + imid * size));
+
+        // three-way comparison
+        if (compare(arr_val, val) > 0)
+            // key is in lower subset
+            return searchsorted(val, arr, count, size, compare, imin, imid - 1);
+        else if (compare(arr_val, val) < 0)
+            // key is in upper subset
+            return searchsorted(val, arr, count, size, compare, imid + 1, imax);
+        else
+            // key has been found
+            return imid;
+    }
+}
+
+
+int grid_index(int i, int j, int k, int dim, index_type type)
+{
+    int ind = -1;
+
+    switch (type) {
+    case INDEX_PADDED:
+        ind = k + (2 * (dim / 2 + 1)) * (j + dim * i);
+        break;
+    case INDEX_REAL:
+        ind = k + dim * (j + dim * i);
+        break;
+    case INDEX_COMPLEX_HERM:
+        ind = k + (dim / 2 + 1) * (j + dim * i);
+        break;
+    default:
+        endrun(1, "Unknown indexing type in `grid_index`.");
+        break;
+    }
+
+    return ind;
+}
+
+
+static void populate_grids()
+{
+    // TODO(smutch): Is this stored somewhere globally?
+    int uvbg_dim = All.UVBGdim;
+    int nranks = -1, this_rank = -1;
+    MPI_Comm_size(MPI_COMM_WORLD, &nranks);
+    MPI_Comm_rank(MPI_COMM_WORLD, &this_rank);
+    ptrdiff_t *slab_nix = UVBGgrids.slab_nix;
+    ptrdiff_t *slab_ix_start = UVBGgrids.slab_ix_start;
+
+    // create buffers on each rank which is as large as the largest LOGICAL allocation on any single rank
+    int buffer_size = 0;
+    for (int ii = 0; ii < nranks; ii++)
+        if (slab_nix[ii] > buffer_size)
+            buffer_size = (int)slab_nix[ii];
+
+    buffer_size *= uvbg_dim * uvbg_dim;
+    float *buffer_mass = fftwf_alloc_real((size_t)buffer_size);
+    float *buffer_stars_slab = fftwf_alloc_real((size_t)buffer_size);
+    float *buffer_sfr = fftwf_alloc_real((size_t)buffer_size);
+
+    //RegionInd no longer global, allocate array for slab decomposition (Jdavies)
+    int *UVRegionInd = mymalloc("UVRegionInd",sizeof(int) * PartManager->NumPart);
+
+    // I am going to reuse the RegionInd member which is from petapm.  I
+    // *think* this is ok as we are doing this after the grav calculations.
+    // This is a potentially stupid way to do things anyway and will most
+    // definitely need to be changed! There is no way we should have to search
+    // all of the particles to find out what slab they sit on, and then loop
+    // through all particles again n_slab times!
+    double box_size = All.BoxSize;
+
+    #pragma omp parallel for
+    for(int ii = 0; ii < PartManager->NumPart; ii++) {
+        if((!P[ii].IsGarbage) && (!P[ii].Swallowed) && (P[ii].Type < 5)) {
+            ptrdiff_t ix = pos_to_ngp(P[ii].Pos[0], box_size, uvbg_dim);
+            UVRegionInd[ii] = searchsorted(&ix, slab_ix_start, nranks, sizeof(ptrdiff_t), compare_ptrdiff, -1, -1);
```

Comment:
What is this code doing? Is it finding the nearest slab?

Reply:
Yes, it fills out an array with which slab each particle is on by comparing the particle x coordinate with the array of slab start coordinates. As the comment says above it there's probably a better way to do the decomposition but I haven't looked at this too closely yet, especially since I was planning to completely change the decomposition later on when switching to pfft.

Reply:
I see! I'm happy with doing this (and maybe we change when moving to pfft), but a comment would be good.